### PR TITLE
identify: filter nat64 well-known prefix ipv6 addresses

### LINKED ISSUE #2349: Handle dial prioritisation for well known prefix IPv4 embedded IPv6 addresses
When implementing happy eyeballs logic to prefer QUIC IPv6 over QUIC IPv4 addresses, I've observed that a lot of the dial failures to IPv6 addresses are for addresses of this form `64:ff9b::192.0.2.33`. These are https://datatracker.ietf.org/doc/html/rfc6052#section-3.1 addresses. 

In most cases these nodes also report `192.0.2.33` as their address and the subsequent dial to that address succeeds. 
Can we just skip the dial to `64:ff9b::192.0.2.33`? 

If we cannot skip them, we should probably still rank them the same as IPv4 addresses and not prefer them to other IPv4 addresses.

### PR DESCRIPTION
**Pull request by sukunrt**
fixes #2349 

For details see https://github.com/libp2p/go-libp2p/issues/2349#issuecomment-1608033872

### CODE REVIEW

File: p2p/protocol/identify/obsaddr.go [hunk 1]
```diff
@@ -435,6 +441,20 @@ func shouldRecordObservation(host addrsProvider, network listenAddrsProvider, co
 	return true
 }
 
+// isNAT64IPv4ConvertedIPv6Addr returns whether addr is an IPv6 address that begins with
+// the well known prefix "64:ff9b" used for NAT64 Translation
+// see RFC 6052
+func isNAT64IPv4ConvertedIPv6Addr(addr ma.Multiaddr) bool {
```
Comment (marten-seemann) [inline p2p/protocol/identify/obsaddr.go:447]: Should this be part of the `manet` package?

File: p2p/protocol/identify/obsaddr.go [hunk 2]
```diff
@@ -435,6 +441,20 @@ func shouldRecordObservation(host addrsProvider, network listenAddrsProvider, co
...
+	ip, err := addr.ValueForProtocol(ma.P_IP6)
+	if err != nil {
+		return false
+	}
+	if strings.HasPrefix(ip, "64:ff9b::") {
```
Comment (marten-seemann) [inline p2p/protocol/identify/obsaddr.go:452]: Is string matching the best way to do this? Would be nicer to match on the underlying bytes (and probably also causes fewer allocs?).

File: p2p/protocol/identify/obsaddr.go (hunk 1 above)
Comment (sukunrt) [reply p2p/protocol/identify/obsaddr.go:447]: will move.

File: p2p/protocol/identify/obsaddr.go (hunk 2 above)
Comment (marten-seemann) [reply p2p/protocol/identify/obsaddr.go:452]: Ok, let's keep it for now, but we should keep this is mind for our multiaddrs API redesign. It's absurd that we return IPs as strings.

**Review by Jorropo — APPROVED** [review]
NAT64 adresses shouldn't be advertised, this makes sense to me.

**Comment by sukunrt** [comment]
blocked on https://github.com/multiformats/go-multiaddr/pull/205

**Comment by sukunrt** [comment]
updated.

**Review by marten-seemann — APPROVED** [review]

**Pull request merged**